twill-yarn/src/main/hadoop21/org/apache/twill/internal/yarn/Hadoop21YarnAppClient.java

```diff
@@ -197,6 +213,7 @@ public YarnApplicationReport getReport() {
     public void cancel() {
       try {
         yarnClient.killApplication(appId);
+        yarnClient.stop();
```

Comment:
You have to keep this `yarnClient` for killing though since it is binding to the user who launched the app.

You probably have to make `ProcessController` `AutoCloseable` so that on application completion, you can close it from the `YarnTwillController`.